proposals/2024-09-25_metadata-labels.md

```diff
@@ -0,0 +1,287 @@
+* It should be a small amount of effort to hide these labels.
+
+### Omit __unit__ label from counting series for summaries and histograms
+
+As explained in [complex values](#complex-metrics-without-native-type-eg-classic-histogram-and-summary) `_count` and `_bucket` has a unit of `observations` or `1`, not the histogram unit that relates to boundaries (`le` label) and `_sum` value. We could drop `__unit__` label for the `_count` and `_bucket` series (and similar for summaries).
+
+However, this would re-introduce part of the problem we are trying to solve. Histogram buckets' `le` label's value is in the units of the metric. If the unit is not included in the labels of the histogram bucket series, series with `le=1000` in seconds, for example, would collide with series with `le=1000` in milliseconds. We could include the unit in the `le` label (e.g. `le=1000s`), but that would be much more disruptive to users without much additional benefit.
+
+The `_count` series of histograms and summaries could omit the `__unit__` label without this consequence, since the count does not have any relation to the unit. This proposal includes the `__unit__` label for consistency so that users can always query for metrics that have a specific unit.
+
+### Special __type__ values for some classic histogram and summary series
+
+We also discussed special [types](https://github.com/prometheus/proposals/pull/39#discussion_r1927374088) and could put to histogram e.g. _count series.
+
+Rejected due to complexity.
+
+### __type__ and __unit__ from client libraries
+
+There is an unwritten rule, claiming that client libraries are not allowed to expose special labels starting with `__` e.g. see [the client_golang check](https://github.com/prometheus/client_golang/blob/34eaefd8a58ff01b243b36a369615859932de9d8/prometheus/labels.go#L187). This combined
+
+One obvious extension of this proposal would be for Prometheus clients to start sending `__type__` and `__unit__` labels with the exposition format. This would:
+
+* Allow mixing metrics with the same name, but different types and units in the same endpoint (an explicit non-goal of this proposal).
+
+This is excluded from this proposal because:
+
+* OpenTelemetry users can use [Views](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#view) to resolve collisions. That should be "good enough".
+* Doing this would require some changes to client libraries, which is significantly more work, and is harder to experiment with.
+* There can be conflicts between `# TYPE` and `# UNIT` metadata for a metric, and `__type__` and `__unit__` labels. This adds complexity to understanding the exposition format, and requires establishing rules for dealing with conflicts.
+
+## Open Questions
+
+1. Should we actually clearly [define PromQL supported types and units in this proposal](#more-strict-unit-and-type-value-definition)? Is there a way to delegate it to another one given a scope?
```

Comment:
I think all parsers included in Prometheus already do this validation, no? I'm not 100% sure, but I think we can't expose metric types that Prometheus doesn't support.. it will fail the scrape with errors

Reply:
TODO: Explain the problem more.

Reply:
Explained in ### More strict unit and type value definition

It depends on your ingestion format. Especially with this rule we mention in `Prometheus Server Ingestion` section where we accept type and unit label if metadata is not set (or metadata support is missing) - it's likely easy to inject `SeCoNdS` value.